src/Cardano/Wallet/BlockSyncer.hs

```diff
@@ -8,100 +9,76 @@
-    -- ^ a way to get a new block
-    -> (Block -> IO ())
-    -- ^ action taken on a new block
+-- | Every interval @delay@, fetches some data from a given source, and call
+-- an action for each elements retrieved.
+tick
+    :: forall st m b. (MonadIO m)
+    => (st -> m ([b], st))
+    -- ^ A way to get a new elements
+    -> (b -> m ())
+    -- ^ Action to be taken on new elements
     -> Millisecond
     -- ^ tick time
-    -> BlockHeadersConsumed st
-    -> IO ()
-tickingFunction getNextBlocks action tickTime = go
-    where
-      go :: BlockHeadersConsumed st -> IO ()
-      go (BlockHeadersConsumed headersConsumed st) = do
-          (st', blocksDownloaded) <- getNextBlocks st
-          let blocksToProcess = filter
-                  (checkIfAlreadyConsumed headersConsumed)
-                  (L.nub blocksDownloaded)
-          mapM_ action blocksToProcess
-          threadDelay $ (fromIntegral . toMicroseconds) tickTime
-          let headersConsumed' = take storingLimit
-                  $ map header blocksToProcess ++ headersConsumed
-          go $ BlockHeadersConsumed headersConsumed' st'
-
-      checkIfAlreadyConsumed
-          :: [BlockHeader]
-          -> Block
-          -> Bool
-      checkIfAlreadyConsumed consumedHeaders (Block theHeader _) =
-          theHeader `L.notElem` consumedHeaders
-
--- | Start the chain producer process, consuming blocks by printing their slot.
-startBlockSyncer :: Text -> Int -> IO ()
-startBlockSyncer networkName port = do
-    network <- newNetworkLayer networkName port
+    -> st
+    -> m ()
+tick next action delay !st = do
+    (bs, !st') <- next st
+    mapM_ action bs
+    liftIO $ threadDelay $ (fromIntegral . toMicroseconds) delay
+    tick next action delay st'
 
-    let interval = 20000 :: Millisecond
-
-        produceBlocks :: SlotId -> IO (SlotId, [Block])
-        produceBlocks start = do
-            res <- runExceptT $ nextBlocks network start
-            case res of
-                Left err -> die $ fmt $ "Chain producer error: "+||err||+""
-                Right [] -> pure (start, [])
-                Right blocks ->
-                    -- fixme: there are more blocks available, so we need not
-                    -- wait for an interval to pass before getting more blocks.
-                    let start' = succ . slotId . header . last $ blocks
-                    in pure (start', blocks)
-
-        logBlock :: Block -> IO ()
-        logBlock block = putStrLn msg
-            where
-                msg = fmt $ "Received block "+||slotId h||+""
-                h = header block
-
-    tickingFunction produceBlocks logBlock interval $
-        BlockHeadersConsumed [] (SlotId 0 0)
+-- | Retrieve blocks from a chain producer and execute some given action for
+-- each block.
+listen
+    :: forall e0 e1. (Show e0)
+    => NetworkLayer IO e0 e1
+    -> (Block -> IO ())
+    -> IO ()
+listen network action = do
+    tick getNextBlocks action 5000 (SlotId 0 0)
```

Comment:
definitely replacing `BlockHeadersConsumed` with `SlotId` will make it easier to follow in the logs!!!